### PR TITLE
MANA Support

### PR DESCRIPTION
**Pull request by karya0**
This PR allows us to configure DMTCP with MANA-specific restart plugins:
``` 
   ./configure --with-mana-helper-dir=../restart_plugin
```
Here, `..\restart_plugin` directory contains plugin/helper files that are linked into dmtcp_restart and mtcp_restart binaries.

### CODE REVIEW

File: src/dmtcp_launch.cpp [hunk 1]
```diff
@@ -146,6 +147,8 @@ static const char *theUsage =
   "              Skip NOTE messages; if given twice, also skip WARNINGs\n"
   "  --coord-logfile PATH (environment variable DMTCP_COORD_LOG_FILENAME\n"
   "              Coordinator will dump its logs to the given file\n"
+  "  --mpi\n"
+  "              Set it to enable MPI-specific flags.\n"
```
Comment (gc00) [inline src/dmtcp_launch.cpp:151]: It his is `to enable MPI-specific flags`, then what are the other MPI-specific flags?  Or did you want to say `to enable MPI/MANA` ?

Or is this intended as:  `Set it to enable '--mpi' flag for MPI/MANA` ?

Or is it:  `Set it to enable 'dmtcp_launch --mpi'` ?

File: src/dmtcp_launch.cpp [hunk 2]
```diff
@@ -40,6 +40,7 @@ static int testJava(const char **argv);
 static bool testSetuid(const char *filename);
```
Comment (gc00) [inline src/dmtcp_launch.cpp:40]: Please modify the commit message.  I think you intended:
`Added --mpi flag to dmtcp_launch.`

File: src/mtcp/mtcp_restart.c [hunk 3]
```diff
@@ -161,6 +157,9 @@ main(int argc, char *argv[], char **environ)
     if (mtcp_strcmp(argv[0], "--use-gdb") == 0) {
       rinfo.use_gdb = 1;
       shift;
+    } else if (mtcp_strcmp(argv[0], "--mpi") == 0) {
+      runMpiProxy = 1;
```
Comment (gc00) [inline src/mtcp/mtcp_restart.c:161]: @rohgarg had chose `runMpiProxy` when we were still thinking about using a proxy approach for MANA.  In the end, MANA switched to using a split process approach.

MANA still creates a proxy child process on launch, but it then "copies the bits" from the child to the lower half of the parent process.

Can we change `runMpiProxy` perhaps to `launchMPI` ?  The name is motivated by the fact that DMTCP will then launch a trivial MPI application in a child process, which is destined to eventually become the lower half.

@rohgarg, what's your opinion?

**Review by gc00 — APPROVED** [review]
I'm approving it in advance for efficiency.

Please see the inline comments.

In particular, one of the inline comments notes that `runMpiProxy` reflects an obsolete design that MANA no longer uses.  I'm recommending `launchMPI`.

File: src/dmtcp_restart.cpp [hunk 4]
```diff
@@ -147,265 +149,219 @@ static void runMtcpRestart(int fd, ProcessInfo *pInfo);
...
+    dmtcp_dlsym_lib_fnc_offset("linux-vdso", "__vdso_getcpu");
+  uint64_t gettimeofday_offset =
+    dmtcp_dlsym_lib_fnc_offset("linux-vdso", "__vdso_gettimeofday");
+  uint64_t time_offset =
+    dmtcp_dlsym_lib_fnc_offset("linux-vdso", "__vdso_time");
+  JWARNING(!_pInfo.vdsoOffsetMismatch(clock_gettime_offset, getcpu_offset,
+                                      gettimeofday_offset, time_offset))
+    .Text("The vDSO section on the current system is different than"
+          " the host where the checkpoint image was generated. "
+          "Restart may fail if the program calls a function in to"
```
[the remark this answers is not in the document]
Comment (karya0) [reply src/dmtcp_restart.cpp:177]: Fixed.

File: src/dmtcp_restart.cpp [hunk 5]
```diff
@@ -147,265 +149,219 @@ static void runMtcpRestart(int fd, ProcessInfo *pInfo);
...
+          " vDSO, like, gettimeofday(), clock_gettime(), etc.");
```
[the remark this answers is not in the document]
Comment (karya0) [reply src/dmtcp_restart.cpp:178]: Fixed

File: src/dmtcp_launch.cpp (hunk 1 above)
Comment (karya0) [reply src/dmtcp_launch.cpp:151]: Fixed.

File: src/dmtcp_launch.cpp (hunk 2 above)
Comment (karya0) [reply src/dmtcp_launch.cpp:40]: Fixed.

File: src/mtcp/mtcp_restart.c (hunk 3 above)
Comment (karya0) [reply src/mtcp/mtcp_restart.c:161]: Changed to mpiMode.

**Pull request merged**